nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java

```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
+
+      String[] antecedent =
+          getAntecedent(clause, tree, analysis, state, callee, node.getArguments().size());
+      String consequent = getConsequent(clause, tree, analysis, state, callee);
+
+      // Find a single value constraint that is not already known. If more than one arguments with
+      // unknown
+      // nullness affects the method's result, then ignore this clause.
+      Node arg = null;
+      Nullness argAntecedentNullness = null;
+      boolean supported =
+          true; // Set to false if the rule is detected to be one we don't yet support
+
+      for (int i = 0; i < antecedent.length; ++i) {
+        String valueConstraint = antecedent[i].trim();
+        if (valueConstraint.equals("_")) {
+          continue;
+        } else if (valueConstraint.equals("false") || valueConstraint.equals("true")) {
+          // We handle boolean constraints in the case that the boolean argument is the result
+          // of a null or not-null check. For example,
+          // '@Contract("true -> true") boolean func(boolean v)'
+          // called with 'func(obj == null)'
+          // can be interpreted as equivalent to
+          // '@Contract("null -> true") boolean func(@Nullable Object v)'
+          // called with 'func(obj)'
+          // This path unwraps null reference equality and inequality checks
+          // to pass the target (in the above example, 'obj') as arg.
+          Node argument = node.getArgument(i);
+          // isNullTarget is the variable side of a null check. For example, both 'e == null'
+          // and 'null == e' would return the node representing 'e'.
+          Optional<Node> isNullTarget = argument.accept(NullEqualityVisitor.IS_NULL, inputs);
+          // notNullTarget is the variable side of a not-null check. For example, both 'e != null'
+          // and 'null != e' would return the node representing 'e'.
+          Optional<Node> notNullTarget = argument.accept(NullEqualityVisitor.NOT_NULL, inputs);
+          // It is possible for at most one of isNullTarget and notNullTarget to be present.
+          Node nullTestTarget = isNullTarget.orElse(notNullTarget.orElse(null));
+          if (nullTestTarget == null) {
             supported = false;
             break;
-          } else if (valueConstraint.equals("!null")
-              && inputs.valueOfSubNode(node.getArgument(i)).equals(Nullness.NONNULL)) {
-            // We already know this argument can't be null, so we can treat it as not part of the
-            // clause
-            // for the purpose of deciding the non-nullness of the other arguments.
+          }
+          // isNullTarget is equivalent to 'null ->' while notNullTarget is equivalent
+          // to '!null ->'. However, the valueConstraint may reverse the check.
+          boolean inverted = isNullTarget.isPresent() == valueConstraint.equals("false");
+          Nullness antecedentNullness = inverted ? Nullness.NONNULL : Nullness.NULL;
```

Comment:
Writing the steps out to simplify from the inputs, to expectations (sorry if the steps seems simple, just trying to make sure I understand the question :-)

`notNullTarget.present() == true` means we have an argument of the form `obj != null`
`valueConstraint.equals("true")` means the contract includes `true -> [result]`
`inverted` is set as the result of `isNullTarget.isPresent() == valueConstraint.equals("false")` which evaluates to
```java
// isNullTarget.isPresent() must be false when notNullTarget.present()
false == valueConstraint.equals("false")

// valueConstraint.equals("true") means valueConstraint.equals("false") must be false
false == false

// Which evaluates to
true

// since inverted is true, the following evaluates to `NONNULL`
Nullness antecedentNullness = inverted ? Nullness.NONNULL : Nullness.NULL;
```

I think it's a bit more confusingly written after I swapped the old implementation out due to `Optional.isEmpty` not existing in jdk8.

I've updated this section with a truth table because boolean logic is difficult to reason about.

Reply:
I agree the more explicit logic is more readable and that there shouldn't be a significant performance cost

Reply:
Good idea, I've updated the implementation with more explicit logic. I tried extracting the logic to a standalone function, but found the indirection made things a bit harder to follow (but I defer to your judgement if you'd prefer a separate function).